# 05 — Classical Baseline (Binary MNIST 0 vs 1, 4 PCA Features)

**Purpose.** Provide a classical baseline for RQ5 (how does the hybrid
quantum-classical model compare to an equivalent classical model?). The
archived `02_classical_baseline.ipynb` trained on the *wrong* dataset
(full 10-class raw-pixel MNIST) and got 12% accuracy -- not usable for
this comparison. This notebook trains a lightweight feedforward network
on the *identical* binary 0-vs-1 / 4-PCA-feature data the VQC uses, with
the same train/val/test split methodology, so the comparison in RQ5 is
fair (same task, same features, same evaluation protocol).

Trained across the same 8 seeds planned for the VQC sweep (`04E`), since
classical training here is fast (seconds, not hours) -- this produces the
full statistical comparison set in one pass.

In [1]:
from pathlib import Path
import sys

import numpy as np
import pandas as pd
import torch
import torch.nn as nn
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

PROJECT_ROOT = Path.cwd().parent

if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))

from src.utils.seed import set_seed
from src.classical.baseline import ClassicalBaseline
from src.training.trainer import Trainer

print("Project root:", PROJECT_ROOT)

Project root: C:\Work\Quantum-Adversarial-Robustness


## Data

Same binary (0 vs 1) / 4-PCA-feature dataset used by the VQC pipeline.

In [2]:
X_train_full = np.load(PROJECT_ROOT / "data" / "binary" / "X_train.npy")
y_train_full = np.load(PROJECT_ROOT / "data" / "binary" / "y_train.npy")
X_test = np.load(PROJECT_ROOT / "data" / "binary" / "X_test.npy")
y_test = np.load(PROJECT_ROOT / "data" / "binary" / "y_test.npy")

print("X_train_full:", X_train_full.shape, " X_test:", X_test.shape)

X_train_full: (11824, 4)  X_test: (2956, 4)


## Multi-seed training

Same train/val split methodology as the VQC pipeline (80/20 stratified),
`StandardScaler` fit on train only, `Trainer` (the same class used for
the VQC) with `BCEWithLogitsLoss` + Adam. 30 epochs is generous given how
quickly this converges -- `Trainer` keeps the best-val-loss checkpoint.

In [3]:
SEEDS = [42, 43, 44, 45, 46, 47, 48, 49]

MODELS_DIR = PROJECT_ROOT / "results" / "models"
CLASSICAL_DIR = PROJECT_ROOT / "results" / "classical"
MODELS_DIR.mkdir(parents=True, exist_ok=True)
CLASSICAL_DIR.mkdir(parents=True, exist_ok=True)

all_results = []

for seed in SEEDS:
    set_seed(seed)

    X_train, X_val, y_train, y_val = train_test_split(
        X_train_full, y_train_full,
        test_size=0.2, stratify=y_train_full, random_state=seed,
    )

    scaler = StandardScaler().fit(X_train)
    X_train_s = scaler.transform(X_train)
    X_val_s = scaler.transform(X_val)
    X_test_s = scaler.transform(X_test)

    def to_tensors(X, y):
        return (
            torch.tensor(X, dtype=torch.float32),
            torch.tensor(y, dtype=torch.float32).reshape(-1, 1),
        )

    X_train_t, y_train_t = to_tensors(X_train_s, y_train)
    X_val_t, y_val_t = to_tensors(X_val_s, y_val)
    X_test_t, y_test_t = to_tensors(X_test_s, y_test)

    train_loader = torch.utils.data.DataLoader(
        torch.utils.data.TensorDataset(X_train_t, y_train_t),
        batch_size=32, shuffle=True,
    )
    val_loader = torch.utils.data.DataLoader(
        torch.utils.data.TensorDataset(X_val_t, y_val_t),
        batch_size=32, shuffle=False,
    )
    test_loader = torch.utils.data.DataLoader(
        torch.utils.data.TensorDataset(X_test_t, y_test_t),
        batch_size=32, shuffle=False,
    )

    model = ClassicalBaseline(in_features=4, hidden_dim=8)
    optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
    criterion = nn.BCEWithLogitsLoss()

    trainer = Trainer(model, optimizer, criterion)
    trainer.fit(train_loader, val_loader, epochs=30)

    test_loss, test_metrics = trainer.evaluate(test_loader)

    n_params = sum(p.numel() for p in model.parameters())

    checkpoint_path = MODELS_DIR / f"classical_baseline_seed{seed}.pt"
    torch.save(model.state_dict(), checkpoint_path)

    result = {
        "seed": seed,
        "n_params": n_params,
        "test_loss": test_loss,
        "accuracy": test_metrics["accuracy"],
        "precision": test_metrics["precision"],
        "recall": test_metrics["recall"],
        "f1": test_metrics["f1"],
    }
    all_results.append(result)
    print(result)

results_df = pd.DataFrame(all_results)
results_df.to_csv(CLASSICAL_DIR / "classical_baseline_all_seeds.csv", index=False)
print()
print(results_df)

Epoch 001/030 | Train Loss: 0.0802 | Val Loss: 0.0088 | Train Acc: 0.9969 | Val Acc: 0.9966


Epoch 002/030 | Train Loss: 0.0083 | Val Loss: 0.0075 | Train Acc: 0.9976 | Val Acc: 0.9975


Epoch 003/030 | Train Loss: 0.0073 | Val Loss: 0.0073 | Train Acc: 0.9980 | Val Acc: 0.9970


Epoch 004/030 | Train Loss: 0.0065 | Val Loss: 0.0068 | Train Acc: 0.9982 | Val Acc: 0.9975


Epoch 005/030 | Train Loss: 0.0061 | Val Loss: 0.0071 | Train Acc: 0.9981 | Val Acc: 0.9970


Epoch 006/030 | Train Loss: 0.0058 | Val Loss: 0.0074 | Train Acc: 0.9983 | Val Acc: 0.9979


Epoch 007/030 | Train Loss: 0.0055 | Val Loss: 0.0068 | Train Acc: 0.9982 | Val Acc: 0.9975


Epoch 008/030 | Train Loss: 0.0045 | Val Loss: 0.0067 | Train Acc: 0.9986 | Val Acc: 0.9979


Epoch 009/030 | Train Loss: 0.0051 | Val Loss: 0.0068 | Train Acc: 0.9986 | Val Acc: 0.9979


Epoch 010/030 | Train Loss: 0.0050 | Val Loss: 0.0070 | Train Acc: 0.9983 | Val Acc: 0.9979


Epoch 011/030 | Train Loss: 0.0050 | Val Loss: 0.0069 | Train Acc: 0.9986 | Val Acc: 0.9979


Epoch 012/030 | Train Loss: 0.0046 | Val Loss: 0.0069 | Train Acc: 0.9982 | Val Acc: 0.9975


Epoch 013/030 | Train Loss: 0.0047 | Val Loss: 0.0071 | Train Acc: 0.9983 | Val Acc: 0.9975


Epoch 014/030 | Train Loss: 0.0051 | Val Loss: 0.0064 | Train Acc: 0.9988 | Val Acc: 0.9983


Epoch 015/030 | Train Loss: 0.0044 | Val Loss: 0.0066 | Train Acc: 0.9987 | Val Acc: 0.9983


Epoch 016/030 | Train Loss: 0.0046 | Val Loss: 0.0064 | Train Acc: 0.9986 | Val Acc: 0.9983


Epoch 017/030 | Train Loss: 0.0044 | Val Loss: 0.0064 | Train Acc: 0.9987 | Val Acc: 0.9979


Epoch 018/030 | Train Loss: 0.0044 | Val Loss: 0.0064 | Train Acc: 0.9987 | Val Acc: 0.9979


Epoch 019/030 | Train Loss: 0.0044 | Val Loss: 0.0064 | Train Acc: 0.9985 | Val Acc: 0.9979


Epoch 020/030 | Train Loss: 0.0043 | Val Loss: 0.0064 | Train Acc: 0.9989 | Val Acc: 0.9983


Epoch 021/030 | Train Loss: 0.0041 | Val Loss: 0.0065 | Train Acc: 0.9987 | Val Acc: 0.9979


Epoch 022/030 | Train Loss: 0.0042 | Val Loss: 0.0064 | Train Acc: 0.9988 | Val Acc: 0.9983


Epoch 023/030 | Train Loss: 0.0044 | Val Loss: 0.0067 | Train Acc: 0.9988 | Val Acc: 0.9979


Epoch 024/030 | Train Loss: 0.0043 | Val Loss: 0.0064 | Train Acc: 0.9989 | Val Acc: 0.9983


Epoch 025/030 | Train Loss: 0.0039 | Val Loss: 0.0069 | Train Acc: 0.9987 | Val Acc: 0.9983


Epoch 026/030 | Train Loss: 0.0041 | Val Loss: 0.0066 | Train Acc: 0.9988 | Val Acc: 0.9979


Epoch 027/030 | Train Loss: 0.0044 | Val Loss: 0.0070 | Train Acc: 0.9988 | Val Acc: 0.9979


Epoch 028/030 | Train Loss: 0.0040 | Val Loss: 0.0068 | Train Acc: 0.9986 | Val Acc: 0.9979


Epoch 029/030 | Train Loss: 0.0045 | Val Loss: 0.0064 | Train Acc: 0.9987 | Val Acc: 0.9983


Epoch 030/030 | Train Loss: 0.0042 | Val Loss: 0.0066 | Train Acc: 0.9989 | Val Acc: 0.9983
{'seed': 42, 'n_params': 49, 'test_loss': 0.003502597129624589, 'accuracy': 0.9986468200270636, 'precision': 0.9987301587301587, 'recall': 0.9987301587301587, 'f1': 0.9987301587301587}


Epoch 001/030 | Train Loss: 0.1062 | Val Loss: 0.0096 | Train Acc: 0.9970 | Val Acc: 0.9983


Epoch 002/030 | Train Loss: 0.0085 | Val Loss: 0.0070 | Train Acc: 0.9977 | Val Acc: 0.9983


Epoch 003/030 | Train Loss: 0.0067 | Val Loss: 0.0059 | Train Acc: 0.9980 | Val Acc: 0.9983


Epoch 004/030 | Train Loss: 0.0059 | Val Loss: 0.0062 | Train Acc: 0.9982 | Val Acc: 0.9983


Epoch 005/030 | Train Loss: 0.0056 | Val Loss: 0.0058 | Train Acc: 0.9984 | Val Acc: 0.9983


Epoch 006/030 | Train Loss: 0.0055 | Val Loss: 0.0061 | Train Acc: 0.9986 | Val Acc: 0.9983


Epoch 007/030 | Train Loss: 0.0054 | Val Loss: 0.0069 | Train Acc: 0.9985 | Val Acc: 0.9979


Epoch 008/030 | Train Loss: 0.0052 | Val Loss: 0.0059 | Train Acc: 0.9985 | Val Acc: 0.9983


Epoch 009/030 | Train Loss: 0.0051 | Val Loss: 0.0065 | Train Acc: 0.9986 | Val Acc: 0.9979


Epoch 010/030 | Train Loss: 0.0047 | Val Loss: 0.0059 | Train Acc: 0.9985 | Val Acc: 0.9983


Epoch 011/030 | Train Loss: 0.0047 | Val Loss: 0.0066 | Train Acc: 0.9988 | Val Acc: 0.9983


Epoch 012/030 | Train Loss: 0.0046 | Val Loss: 0.0059 | Train Acc: 0.9986 | Val Acc: 0.9987


Epoch 013/030 | Train Loss: 0.0046 | Val Loss: 0.0064 | Train Acc: 0.9988 | Val Acc: 0.9983


Epoch 014/030 | Train Loss: 0.0045 | Val Loss: 0.0059 | Train Acc: 0.9989 | Val Acc: 0.9987


Epoch 015/030 | Train Loss: 0.0044 | Val Loss: 0.0059 | Train Acc: 0.9989 | Val Acc: 0.9987


Epoch 016/030 | Train Loss: 0.0043 | Val Loss: 0.0059 | Train Acc: 0.9988 | Val Acc: 0.9987


Epoch 017/030 | Train Loss: 0.0042 | Val Loss: 0.0066 | Train Acc: 0.9989 | Val Acc: 0.9983


Epoch 018/030 | Train Loss: 0.0043 | Val Loss: 0.0059 | Train Acc: 0.9992 | Val Acc: 0.9987


Epoch 019/030 | Train Loss: 0.0042 | Val Loss: 0.0056 | Train Acc: 0.9987 | Val Acc: 0.9987


Epoch 020/030 | Train Loss: 0.0041 | Val Loss: 0.0060 | Train Acc: 0.9992 | Val Acc: 0.9987


Epoch 021/030 | Train Loss: 0.0042 | Val Loss: 0.0061 | Train Acc: 0.9988 | Val Acc: 0.9979


Epoch 022/030 | Train Loss: 0.0040 | Val Loss: 0.0055 | Train Acc: 0.9988 | Val Acc: 0.9987


Epoch 023/030 | Train Loss: 0.0040 | Val Loss: 0.0053 | Train Acc: 0.9990 | Val Acc: 0.9987


Epoch 024/030 | Train Loss: 0.0040 | Val Loss: 0.0053 | Train Acc: 0.9990 | Val Acc: 0.9987


Epoch 025/030 | Train Loss: 0.0037 | Val Loss: 0.0056 | Train Acc: 0.9992 | Val Acc: 0.9987


Epoch 026/030 | Train Loss: 0.0039 | Val Loss: 0.0052 | Train Acc: 0.9989 | Val Acc: 0.9987


Epoch 027/030 | Train Loss: 0.0038 | Val Loss: 0.0053 | Train Acc: 0.9990 | Val Acc: 0.9987


Epoch 028/030 | Train Loss: 0.0038 | Val Loss: 0.0054 | Train Acc: 0.9985 | Val Acc: 0.9987


Epoch 029/030 | Train Loss: 0.0038 | Val Loss: 0.0052 | Train Acc: 0.9990 | Val Acc: 0.9987


Epoch 030/030 | Train Loss: 0.0036 | Val Loss: 0.0052 | Train Acc: 0.9990 | Val Acc: 0.9987
{'seed': 43, 'n_params': 49, 'test_loss': 0.0032546780200953802, 'accuracy': 0.9983085250338295, 'precision': 0.9974651457541192, 'recall': 0.9993650793650793, 'f1': 0.9984142086901364}


Epoch 001/030 | Train Loss: 0.1024 | Val Loss: 0.0088 | Train Acc: 0.9973 | Val Acc: 0.9975


Epoch 002/030 | Train Loss: 0.0086 | Val Loss: 0.0077 | Train Acc: 0.9966 | Val Acc: 0.9958


Epoch 003/030 | Train Loss: 0.0079 | Val Loss: 0.0057 | Train Acc: 0.9980 | Val Acc: 0.9983


Epoch 004/030 | Train Loss: 0.0066 | Val Loss: 0.0058 | Train Acc: 0.9970 | Val Acc: 0.9975


Epoch 005/030 | Train Loss: 0.0069 | Val Loss: 0.0048 | Train Acc: 0.9982 | Val Acc: 0.9983


Epoch 006/030 | Train Loss: 0.0064 | Val Loss: 0.0049 | Train Acc: 0.9969 | Val Acc: 0.9975


Epoch 007/030 | Train Loss: 0.0063 | Val Loss: 0.0049 | Train Acc: 0.9987 | Val Acc: 0.9983


Epoch 008/030 | Train Loss: 0.0063 | Val Loss: 0.0055 | Train Acc: 0.9985 | Val Acc: 0.9983


Epoch 009/030 | Train Loss: 0.0060 | Val Loss: 0.0042 | Train Acc: 0.9981 | Val Acc: 0.9983


Epoch 010/030 | Train Loss: 0.0055 | Val Loss: 0.0051 | Train Acc: 0.9971 | Val Acc: 0.9983


Epoch 011/030 | Train Loss: 0.0058 | Val Loss: 0.0046 | Train Acc: 0.9986 | Val Acc: 0.9979


Epoch 012/030 | Train Loss: 0.0053 | Val Loss: 0.0058 | Train Acc: 0.9980 | Val Acc: 0.9983


Epoch 013/030 | Train Loss: 0.0054 | Val Loss: 0.0053 | Train Acc: 0.9986 | Val Acc: 0.9979


Epoch 014/030 | Train Loss: 0.0054 | Val Loss: 0.0057 | Train Acc: 0.9987 | Val Acc: 0.9979


Epoch 015/030 | Train Loss: 0.0053 | Val Loss: 0.0053 | Train Acc: 0.9988 | Val Acc: 0.9979


Epoch 016/030 | Train Loss: 0.0053 | Val Loss: 0.0046 | Train Acc: 0.9985 | Val Acc: 0.9979


Epoch 017/030 | Train Loss: 0.0052 | Val Loss: 0.0052 | Train Acc: 0.9988 | Val Acc: 0.9979


Epoch 018/030 | Train Loss: 0.0053 | Val Loss: 0.0048 | Train Acc: 0.9986 | Val Acc: 0.9979


Epoch 019/030 | Train Loss: 0.0050 | Val Loss: 0.0049 | Train Acc: 0.9987 | Val Acc: 0.9983


Epoch 020/030 | Train Loss: 0.0052 | Val Loss: 0.0050 | Train Acc: 0.9981 | Val Acc: 0.9983


Epoch 021/030 | Train Loss: 0.0049 | Val Loss: 0.0053 | Train Acc: 0.9988 | Val Acc: 0.9979


Epoch 022/030 | Train Loss: 0.0049 | Val Loss: 0.0053 | Train Acc: 0.9985 | Val Acc: 0.9979


Epoch 023/030 | Train Loss: 0.0045 | Val Loss: 0.0053 | Train Acc: 0.9985 | Val Acc: 0.9979


Epoch 024/030 | Train Loss: 0.0046 | Val Loss: 0.0057 | Train Acc: 0.9987 | Val Acc: 0.9979


Epoch 025/030 | Train Loss: 0.0051 | Val Loss: 0.0054 | Train Acc: 0.9985 | Val Acc: 0.9979


Epoch 026/030 | Train Loss: 0.0048 | Val Loss: 0.0053 | Train Acc: 0.9984 | Val Acc: 0.9983


Epoch 027/030 | Train Loss: 0.0047 | Val Loss: 0.0052 | Train Acc: 0.9983 | Val Acc: 0.9983


Epoch 028/030 | Train Loss: 0.0048 | Val Loss: 0.0051 | Train Acc: 0.9981 | Val Acc: 0.9979


Epoch 029/030 | Train Loss: 0.0047 | Val Loss: 0.0049 | Train Acc: 0.9986 | Val Acc: 0.9983


Epoch 030/030 | Train Loss: 0.0046 | Val Loss: 0.0051 | Train Acc: 0.9986 | Val Acc: 0.9983
{'seed': 44, 'n_params': 49, 'test_loss': 0.004048109942173184, 'accuracy': 0.9989851150202977, 'precision': 0.998730964467005, 'recall': 0.9993650793650793, 'f1': 0.999047921294827}


Epoch 001/030 | Train Loss: 0.0888 | Val Loss: 0.0074 | Train Acc: 0.9978 | Val Acc: 0.9979


Epoch 002/030 | Train Loss: 0.0080 | Val Loss: 0.0058 | Train Acc: 0.9977 | Val Acc: 0.9975


Epoch 003/030 | Train Loss: 0.0070 | Val Loss: 0.0049 | Train Acc: 0.9981 | Val Acc: 0.9979


Epoch 004/030 | Train Loss: 0.0068 | Val Loss: 0.0047 | Train Acc: 0.9981 | Val Acc: 0.9979


Epoch 005/030 | Train Loss: 0.0063 | Val Loss: 0.0046 | Train Acc: 0.9983 | Val Acc: 0.9983


Epoch 006/030 | Train Loss: 0.0063 | Val Loss: 0.0044 | Train Acc: 0.9984 | Val Acc: 0.9983


Epoch 007/030 | Train Loss: 0.0062 | Val Loss: 0.0040 | Train Acc: 0.9981 | Val Acc: 0.9983


Epoch 008/030 | Train Loss: 0.0059 | Val Loss: 0.0045 | Train Acc: 0.9982 | Val Acc: 0.9983


Epoch 009/030 | Train Loss: 0.0057 | Val Loss: 0.0044 | Train Acc: 0.9982 | Val Acc: 0.9983


Epoch 010/030 | Train Loss: 0.0060 | Val Loss: 0.0043 | Train Acc: 0.9983 | Val Acc: 0.9983


Epoch 011/030 | Train Loss: 0.0054 | Val Loss: 0.0047 | Train Acc: 0.9980 | Val Acc: 0.9987


Epoch 012/030 | Train Loss: 0.0054 | Val Loss: 0.0041 | Train Acc: 0.9981 | Val Acc: 0.9987


Epoch 013/030 | Train Loss: 0.0053 | Val Loss: 0.0038 | Train Acc: 0.9983 | Val Acc: 0.9987


Epoch 014/030 | Train Loss: 0.0053 | Val Loss: 0.0042 | Train Acc: 0.9982 | Val Acc: 0.9987


Epoch 015/030 | Train Loss: 0.0053 | Val Loss: 0.0039 | Train Acc: 0.9984 | Val Acc: 0.9987


Epoch 016/030 | Train Loss: 0.0055 | Val Loss: 0.0037 | Train Acc: 0.9983 | Val Acc: 0.9987


Epoch 017/030 | Train Loss: 0.0053 | Val Loss: 0.0038 | Train Acc: 0.9984 | Val Acc: 0.9983


Epoch 018/030 | Train Loss: 0.0050 | Val Loss: 0.0051 | Train Acc: 0.9981 | Val Acc: 0.9983


Epoch 019/030 | Train Loss: 0.0052 | Val Loss: 0.0039 | Train Acc: 0.9984 | Val Acc: 0.9983


Epoch 020/030 | Train Loss: 0.0051 | Val Loss: 0.0039 | Train Acc: 0.9984 | Val Acc: 0.9987


Epoch 021/030 | Train Loss: 0.0052 | Val Loss: 0.0037 | Train Acc: 0.9986 | Val Acc: 0.9987


Epoch 022/030 | Train Loss: 0.0050 | Val Loss: 0.0035 | Train Acc: 0.9986 | Val Acc: 0.9987


Epoch 023/030 | Train Loss: 0.0049 | Val Loss: 0.0035 | Train Acc: 0.9987 | Val Acc: 0.9987


Epoch 024/030 | Train Loss: 0.0046 | Val Loss: 0.0043 | Train Acc: 0.9984 | Val Acc: 0.9979


Epoch 025/030 | Train Loss: 0.0051 | Val Loss: 0.0031 | Train Acc: 0.9985 | Val Acc: 0.9987


Epoch 026/030 | Train Loss: 0.0047 | Val Loss: 0.0032 | Train Acc: 0.9985 | Val Acc: 0.9987


Epoch 027/030 | Train Loss: 0.0048 | Val Loss: 0.0033 | Train Acc: 0.9986 | Val Acc: 0.9987


Epoch 028/030 | Train Loss: 0.0048 | Val Loss: 0.0029 | Train Acc: 0.9985 | Val Acc: 0.9983


Epoch 029/030 | Train Loss: 0.0047 | Val Loss: 0.0032 | Train Acc: 0.9986 | Val Acc: 0.9992


Epoch 030/030 | Train Loss: 0.0048 | Val Loss: 0.0029 | Train Acc: 0.9986 | Val Acc: 0.9987
{'seed': 45, 'n_params': 49, 'test_loss': 0.0028732938149516095, 'accuracy': 0.9986468200270636, 'precision': 0.9980976537729866, 'recall': 0.9993650793650793, 'f1': 0.998730964467005}


Epoch 001/030 | Train Loss: 0.1494 | Val Loss: 0.0132 | Train Acc: 0.9966 | Val Acc: 0.9958


Epoch 002/030 | Train Loss: 0.0099 | Val Loss: 0.0094 | Train Acc: 0.9978 | Val Acc: 0.9966


Epoch 003/030 | Train Loss: 0.0074 | Val Loss: 0.0083 | Train Acc: 0.9977 | Val Acc: 0.9962


Epoch 004/030 | Train Loss: 0.0069 | Val Loss: 0.0077 | Train Acc: 0.9982 | Val Acc: 0.9975


Epoch 005/030 | Train Loss: 0.0064 | Val Loss: 0.0069 | Train Acc: 0.9976 | Val Acc: 0.9975


Epoch 006/030 | Train Loss: 0.0063 | Val Loss: 0.0064 | Train Acc: 0.9979 | Val Acc: 0.9975


Epoch 007/030 | Train Loss: 0.0057 | Val Loss: 0.0064 | Train Acc: 0.9979 | Val Acc: 0.9975


Epoch 008/030 | Train Loss: 0.0057 | Val Loss: 0.0065 | Train Acc: 0.9983 | Val Acc: 0.9979


Epoch 009/030 | Train Loss: 0.0057 | Val Loss: 0.0066 | Train Acc: 0.9985 | Val Acc: 0.9979


Epoch 010/030 | Train Loss: 0.0056 | Val Loss: 0.0059 | Train Acc: 0.9985 | Val Acc: 0.9979


Epoch 011/030 | Train Loss: 0.0054 | Val Loss: 0.0067 | Train Acc: 0.9985 | Val Acc: 0.9979


Epoch 012/030 | Train Loss: 0.0052 | Val Loss: 0.0059 | Train Acc: 0.9986 | Val Acc: 0.9983


Epoch 013/030 | Train Loss: 0.0051 | Val Loss: 0.0058 | Train Acc: 0.9985 | Val Acc: 0.9979


Epoch 014/030 | Train Loss: 0.0052 | Val Loss: 0.0057 | Train Acc: 0.9986 | Val Acc: 0.9983


Epoch 015/030 | Train Loss: 0.0049 | Val Loss: 0.0061 | Train Acc: 0.9987 | Val Acc: 0.9987


Epoch 016/030 | Train Loss: 0.0049 | Val Loss: 0.0056 | Train Acc: 0.9984 | Val Acc: 0.9979


Epoch 017/030 | Train Loss: 0.0050 | Val Loss: 0.0057 | Train Acc: 0.9987 | Val Acc: 0.9987


Epoch 018/030 | Train Loss: 0.0049 | Val Loss: 0.0054 | Train Acc: 0.9986 | Val Acc: 0.9983


Epoch 019/030 | Train Loss: 0.0047 | Val Loss: 0.0056 | Train Acc: 0.9982 | Val Acc: 0.9983


Epoch 020/030 | Train Loss: 0.0049 | Val Loss: 0.0051 | Train Acc: 0.9988 | Val Acc: 0.9987


Epoch 021/030 | Train Loss: 0.0047 | Val Loss: 0.0052 | Train Acc: 0.9987 | Val Acc: 0.9987


Epoch 022/030 | Train Loss: 0.0048 | Val Loss: 0.0049 | Train Acc: 0.9987 | Val Acc: 0.9987


Epoch 023/030 | Train Loss: 0.0046 | Val Loss: 0.0053 | Train Acc: 0.9983 | Val Acc: 0.9983


Epoch 024/030 | Train Loss: 0.0047 | Val Loss: 0.0053 | Train Acc: 0.9988 | Val Acc: 0.9987


Epoch 025/030 | Train Loss: 0.0048 | Val Loss: 0.0056 | Train Acc: 0.9987 | Val Acc: 0.9983


Epoch 026/030 | Train Loss: 0.0048 | Val Loss: 0.0054 | Train Acc: 0.9988 | Val Acc: 0.9987


Epoch 027/030 | Train Loss: 0.0043 | Val Loss: 0.0058 | Train Acc: 0.9988 | Val Acc: 0.9983


Epoch 028/030 | Train Loss: 0.0047 | Val Loss: 0.0059 | Train Acc: 0.9988 | Val Acc: 0.9983


Epoch 029/030 | Train Loss: 0.0041 | Val Loss: 0.0078 | Train Acc: 0.9976 | Val Acc: 0.9962


Epoch 030/030 | Train Loss: 0.0050 | Val Loss: 0.0063 | Train Acc: 0.9988 | Val Acc: 0.9983
{'seed': 46, 'n_params': 49, 'test_loss': 0.004020918284653761, 'accuracy': 0.9986468200270636, 'precision': 0.9987301587301587, 'recall': 0.9987301587301587, 'f1': 0.9987301587301587}


Epoch 001/030 | Train Loss: 0.1998 | Val Loss: 0.0695 | Train Acc: 0.9976 | Val Acc: 0.9975


Epoch 002/030 | Train Loss: 0.0451 | Val Loss: 0.0304 | Train Acc: 0.9976 | Val Acc: 0.9979


Epoch 003/030 | Train Loss: 0.0236 | Val Loss: 0.0192 | Train Acc: 0.9978 | Val Acc: 0.9970


Epoch 004/030 | Train Loss: 0.0163 | Val Loss: 0.0144 | Train Acc: 0.9979 | Val Acc: 0.9970


Epoch 005/030 | Train Loss: 0.0128 | Val Loss: 0.0122 | Train Acc: 0.9978 | Val Acc: 0.9975


Epoch 006/030 | Train Loss: 0.0110 | Val Loss: 0.0108 | Train Acc: 0.9978 | Val Acc: 0.9975


Epoch 007/030 | Train Loss: 0.0099 | Val Loss: 0.0102 | Train Acc: 0.9978 | Val Acc: 0.9970


Epoch 008/030 | Train Loss: 0.0092 | Val Loss: 0.0089 | Train Acc: 0.9977 | Val Acc: 0.9979


Epoch 009/030 | Train Loss: 0.0087 | Val Loss: 0.0089 | Train Acc: 0.9976 | Val Acc: 0.9970


Epoch 010/030 | Train Loss: 0.0078 | Val Loss: 0.0102 | Train Acc: 0.9977 | Val Acc: 0.9979


Epoch 011/030 | Train Loss: 0.0080 | Val Loss: 0.0082 | Train Acc: 0.9981 | Val Acc: 0.9966


Epoch 012/030 | Train Loss: 0.0076 | Val Loss: 0.0079 | Train Acc: 0.9980 | Val Acc: 0.9975


Epoch 013/030 | Train Loss: 0.0074 | Val Loss: 0.0089 | Train Acc: 0.9983 | Val Acc: 0.9975


Epoch 014/030 | Train Loss: 0.0071 | Val Loss: 0.0087 | Train Acc: 0.9984 | Val Acc: 0.9975


Epoch 015/030 | Train Loss: 0.0068 | Val Loss: 0.0077 | Train Acc: 0.9981 | Val Acc: 0.9979


Epoch 016/030 | Train Loss: 0.0066 | Val Loss: 0.0076 | Train Acc: 0.9982 | Val Acc: 0.9975


Epoch 017/030 | Train Loss: 0.0066 | Val Loss: 0.0077 | Train Acc: 0.9983 | Val Acc: 0.9975


Epoch 018/030 | Train Loss: 0.0065 | Val Loss: 0.0077 | Train Acc: 0.9983 | Val Acc: 0.9979


Epoch 019/030 | Train Loss: 0.0068 | Val Loss: 0.0074 | Train Acc: 0.9983 | Val Acc: 0.9979


Epoch 020/030 | Train Loss: 0.0067 | Val Loss: 0.0075 | Train Acc: 0.9982 | Val Acc: 0.9975


Epoch 021/030 | Train Loss: 0.0062 | Val Loss: 0.0084 | Train Acc: 0.9985 | Val Acc: 0.9975


Epoch 022/030 | Train Loss: 0.0062 | Val Loss: 0.0076 | Train Acc: 0.9983 | Val Acc: 0.9975


Epoch 023/030 | Train Loss: 0.0061 | Val Loss: 0.0086 | Train Acc: 0.9985 | Val Acc: 0.9975


Epoch 024/030 | Train Loss: 0.0062 | Val Loss: 0.0082 | Train Acc: 0.9982 | Val Acc: 0.9975


Epoch 025/030 | Train Loss: 0.0060 | Val Loss: 0.0079 | Train Acc: 0.9982 | Val Acc: 0.9975


Epoch 026/030 | Train Loss: 0.0061 | Val Loss: 0.0088 | Train Acc: 0.9985 | Val Acc: 0.9979


Epoch 027/030 | Train Loss: 0.0058 | Val Loss: 0.0070 | Train Acc: 0.9982 | Val Acc: 0.9970


Epoch 028/030 | Train Loss: 0.0058 | Val Loss: 0.0079 | Train Acc: 0.9984 | Val Acc: 0.9975


Epoch 029/030 | Train Loss: 0.0059 | Val Loss: 0.0072 | Train Acc: 0.9983 | Val Acc: 0.9970


Epoch 030/030 | Train Loss: 0.0056 | Val Loss: 0.0077 | Train Acc: 0.9983 | Val Acc: 0.9975
{'seed': 47, 'n_params': 49, 'test_loss': 0.0038567342049444123, 'accuracy': 0.9986468200270636, 'precision': 0.9987301587301587, 'recall': 0.9987301587301587, 'f1': 0.9987301587301587}


Epoch 001/030 | Train Loss: 0.1252 | Val Loss: 0.0101 | Train Acc: 0.9968 | Val Acc: 0.9979


Epoch 002/030 | Train Loss: 0.0102 | Val Loss: 0.0067 | Train Acc: 0.9968 | Val Acc: 0.9979


Epoch 003/030 | Train Loss: 0.0091 | Val Loss: 0.0061 | Train Acc: 0.9968 | Val Acc: 0.9979


Epoch 004/030 | Train Loss: 0.0091 | Val Loss: 0.0066 | Train Acc: 0.9970 | Val Acc: 0.9979


Epoch 005/030 | Train Loss: 0.0085 | Val Loss: 0.0054 | Train Acc: 0.9971 | Val Acc: 0.9979


Epoch 006/030 | Train Loss: 0.0085 | Val Loss: 0.0066 | Train Acc: 0.9969 | Val Acc: 0.9979


Epoch 007/030 | Train Loss: 0.0087 | Val Loss: 0.0068 | Train Acc: 0.9969 | Val Acc: 0.9979


Epoch 008/030 | Train Loss: 0.0080 | Val Loss: 0.0077 | Train Acc: 0.9965 | Val Acc: 0.9979


Epoch 009/030 | Train Loss: 0.0084 | Val Loss: 0.0057 | Train Acc: 0.9971 | Val Acc: 0.9979


Epoch 010/030 | Train Loss: 0.0083 | Val Loss: 0.0057 | Train Acc: 0.9973 | Val Acc: 0.9979


Epoch 011/030 | Train Loss: 0.0085 | Val Loss: 0.0055 | Train Acc: 0.9971 | Val Acc: 0.9979


Epoch 012/030 | Train Loss: 0.0087 | Val Loss: 0.0055 | Train Acc: 0.9976 | Val Acc: 0.9979


Epoch 013/030 | Train Loss: 0.0085 | Val Loss: 0.0059 | Train Acc: 0.9974 | Val Acc: 0.9979


Epoch 014/030 | Train Loss: 0.0084 | Val Loss: 0.0070 | Train Acc: 0.9974 | Val Acc: 0.9979


Epoch 015/030 | Train Loss: 0.0082 | Val Loss: 0.0057 | Train Acc: 0.9974 | Val Acc: 0.9979


Epoch 016/030 | Train Loss: 0.0084 | Val Loss: 0.0085 | Train Acc: 0.9966 | Val Acc: 0.9983


Epoch 017/030 | Train Loss: 0.0086 | Val Loss: 0.0064 | Train Acc: 0.9974 | Val Acc: 0.9979


Epoch 018/030 | Train Loss: 0.0085 | Val Loss: 0.0074 | Train Acc: 0.9973 | Val Acc: 0.9979


Epoch 019/030 | Train Loss: 0.0088 | Val Loss: 0.0065 | Train Acc: 0.9970 | Val Acc: 0.9979


Epoch 020/030 | Train Loss: 0.0081 | Val Loss: 0.0059 | Train Acc: 0.9974 | Val Acc: 0.9979


Epoch 021/030 | Train Loss: 0.0084 | Val Loss: 0.0062 | Train Acc: 0.9976 | Val Acc: 0.9979


Epoch 022/030 | Train Loss: 0.0083 | Val Loss: 0.0063 | Train Acc: 0.9976 | Val Acc: 0.9979


Epoch 023/030 | Train Loss: 0.0085 | Val Loss: 0.0065 | Train Acc: 0.9975 | Val Acc: 0.9979


Epoch 024/030 | Train Loss: 0.0081 | Val Loss: 0.0061 | Train Acc: 0.9976 | Val Acc: 0.9979


Epoch 025/030 | Train Loss: 0.0087 | Val Loss: 0.0065 | Train Acc: 0.9974 | Val Acc: 0.9979


Epoch 026/030 | Train Loss: 0.0083 | Val Loss: 0.0055 | Train Acc: 0.9975 | Val Acc: 0.9979


Epoch 027/030 | Train Loss: 0.0081 | Val Loss: 0.0072 | Train Acc: 0.9974 | Val Acc: 0.9979


Epoch 028/030 | Train Loss: 0.0083 | Val Loss: 0.0055 | Train Acc: 0.9977 | Val Acc: 0.9979


Epoch 029/030 | Train Loss: 0.0081 | Val Loss: 0.0066 | Train Acc: 0.9975 | Val Acc: 0.9979


Epoch 030/030 | Train Loss: 0.0082 | Val Loss: 0.0059 | Train Acc: 0.9977 | Val Acc: 0.9979
{'seed': 48, 'n_params': 49, 'test_loss': 0.007804829440522606, 'accuracy': 0.9962787550744249, 'precision': 0.9987244897959183, 'recall': 0.9942857142857143, 'f1': 0.9965001590836781}


Epoch 001/030 | Train Loss: 0.1389 | Val Loss: 0.0113 | Train Acc: 0.9966 | Val Acc: 0.9966


Epoch 002/030 | Train Loss: 0.0101 | Val Loss: 0.0084 | Train Acc: 0.9964 | Val Acc: 0.9958


Epoch 003/030 | Train Loss: 0.0086 | Val Loss: 0.0067 | Train Acc: 0.9967 | Val Acc: 0.9975


Epoch 004/030 | Train Loss: 0.0083 | Val Loss: 0.0053 | Train Acc: 0.9974 | Val Acc: 0.9975


Epoch 005/030 | Train Loss: 0.0079 | Val Loss: 0.0051 | Train Acc: 0.9976 | Val Acc: 0.9979


Epoch 006/030 | Train Loss: 0.0076 | Val Loss: 0.0050 | Train Acc: 0.9976 | Val Acc: 0.9979


Epoch 007/030 | Train Loss: 0.0075 | Val Loss: 0.0054 | Train Acc: 0.9974 | Val Acc: 0.9979


Epoch 008/030 | Train Loss: 0.0069 | Val Loss: 0.0075 | Train Acc: 0.9970 | Val Acc: 0.9970


Epoch 009/030 | Train Loss: 0.0074 | Val Loss: 0.0051 | Train Acc: 0.9978 | Val Acc: 0.9979


Epoch 010/030 | Train Loss: 0.0072 | Val Loss: 0.0055 | Train Acc: 0.9976 | Val Acc: 0.9979


Epoch 011/030 | Train Loss: 0.0072 | Val Loss: 0.0054 | Train Acc: 0.9980 | Val Acc: 0.9979


Epoch 012/030 | Train Loss: 0.0077 | Val Loss: 0.0056 | Train Acc: 0.9974 | Val Acc: 0.9975


Epoch 013/030 | Train Loss: 0.0071 | Val Loss: 0.0049 | Train Acc: 0.9980 | Val Acc: 0.9979


Epoch 014/030 | Train Loss: 0.0069 | Val Loss: 0.0045 | Train Acc: 0.9979 | Val Acc: 0.9979


Epoch 015/030 | Train Loss: 0.0072 | Val Loss: 0.0056 | Train Acc: 0.9978 | Val Acc: 0.9979


Epoch 016/030 | Train Loss: 0.0071 | Val Loss: 0.0043 | Train Acc: 0.9981 | Val Acc: 0.9979


Epoch 017/030 | Train Loss: 0.0068 | Val Loss: 0.0042 | Train Acc: 0.9976 | Val Acc: 0.9979


Epoch 018/030 | Train Loss: 0.0069 | Val Loss: 0.0044 | Train Acc: 0.9984 | Val Acc: 0.9979


Epoch 019/030 | Train Loss: 0.0064 | Val Loss: 0.0062 | Train Acc: 0.9979 | Val Acc: 0.9975


Epoch 020/030 | Train Loss: 0.0064 | Val Loss: 0.0048 | Train Acc: 0.9980 | Val Acc: 0.9979


Epoch 021/030 | Train Loss: 0.0065 | Val Loss: 0.0045 | Train Acc: 0.9981 | Val Acc: 0.9983


Epoch 022/030 | Train Loss: 0.0065 | Val Loss: 0.0048 | Train Acc: 0.9982 | Val Acc: 0.9979


Epoch 023/030 | Train Loss: 0.0063 | Val Loss: 0.0050 | Train Acc: 0.9983 | Val Acc: 0.9983


Epoch 024/030 | Train Loss: 0.0066 | Val Loss: 0.0036 | Train Acc: 0.9982 | Val Acc: 0.9979


Epoch 025/030 | Train Loss: 0.0066 | Val Loss: 0.0042 | Train Acc: 0.9983 | Val Acc: 0.9979


Epoch 026/030 | Train Loss: 0.0065 | Val Loss: 0.0041 | Train Acc: 0.9980 | Val Acc: 0.9983


Epoch 027/030 | Train Loss: 0.0067 | Val Loss: 0.0044 | Train Acc: 0.9984 | Val Acc: 0.9979


Epoch 028/030 | Train Loss: 0.0065 | Val Loss: 0.0052 | Train Acc: 0.9981 | Val Acc: 0.9983


Epoch 029/030 | Train Loss: 0.0063 | Val Loss: 0.0046 | Train Acc: 0.9982 | Val Acc: 0.9979


Epoch 030/030 | Train Loss: 0.0065 | Val Loss: 0.0045 | Train Acc: 0.9983 | Val Acc: 0.9983
{'seed': 49, 'n_params': 49, 'test_loss': 0.004542062712463245, 'accuracy': 0.9979702300405954, 'precision': 0.9987285441830897, 'recall': 0.9974603174603175, 'f1': 0.9980940279542567}

   seed  n_params  test_loss  accuracy  precision    recall        f1
0    42        49   0.003503  0.998647   0.998730  0.998730  0.998730
1    43        49   0.003255  0.998309   0.997465  0.999365  0.998414
2    44        49   0.004048  0.998985   0.998731  0.999365  0.999048
3    45        49   0.002873  0.998647   0.998098  0.999365  0.998731
4    46        49   0.004021  0.998647   0.998730  0.998730  0.998730
5    47        49   0.003857  0.998647   0.998730  0.998730  0.998730
6    48        49   0.007805  0.996279   0.998724  0.994286  0.996500
7    49        49   0.004542  0.997970   0.998729  0.997460  0.998094


## Summary statistics across seeds

In [4]:
summary = results_df[["accuracy", "precision", "recall", "f1"]].agg(["mean", "std"])
print(summary)

summary.to_json(CLASSICAL_DIR / "classical_baseline_summary.json")

      accuracy  precision    recall        f1
mean  0.998266   0.998492  0.998254  0.998372
std   0.000857   0.000470  0.001722  0.000807


**Interpretation.** With `StandardScaler` correctly fit per-split and
`Trainer`'s best-val-loss checkpointing, this lightweight classical
network should land close to the ~99.66% ceiling already established by
plain `LogisticRegression`/`SVC(rbf)` sanity checks in the Phase 0
diagnostics (`04D`) -- this is the number RQ5 will compare the VQC's
~63-64% (Phase 0-accepted `single_z`/`ansatz_reps=4` configuration)
against.